From 0b8d3cb468dbd2188cdd13466674154862d2a07b Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Thu, 27 Feb 2025 14:14:58 +0800
Subject: [PATCH 416/652] fix(dambuf helper):get the mutex while operating on
 the list

Changelogs:
1.INIT_LIST_HEAD(&heap_obj->list) before adding to the list head.
2.Hold the root->lock when performing the uninit.

Signed-off-by: linmin <linmin@eswincomputing.com>
---
 .../dmabuf-heap-import-helper.c               | 29 +++++++++++++++----
 1 file changed, 23 insertions(+), 6 deletions(-)

diff --git a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
index 283fbdfb1c071..0aebc309c3a96 100644
--- a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
+++ b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
@@ -165,9 +165,11 @@ void common_dmabuf_heap_import_uninit(struct heap_root *root)
 {
 	struct heap_mem *h, *tmp;
 
+	mutex_lock(&root->lock);
 	list_for_each_entry_safe(h, tmp, &root->header, list) {
 		common_dmabuf_heap_release(h);
 	}
+	mutex_unlock(&root->lock);
 }
 EXPORT_SYMBOL(common_dmabuf_heap_import_uninit);
 
@@ -234,6 +236,7 @@ static struct heap_mem *dmabuf_heap_import(struct heap_root *root, int fd)
 
     kref_init(&heap_obj->refcount);
 
+    INIT_LIST_HEAD(&heap_obj->list);
     list_add(&heap_obj->list, &root->header);
 
     mutex_unlock(&root->lock);
@@ -308,6 +311,7 @@ static struct heap_mem *dmabuf_heap_import_with_dma_buf_st(struct heap_root *roo
 
     kref_init(&heap_obj->refcount);
 
+    INIT_LIST_HEAD(&heap_obj->list);
     list_add(&heap_obj->list, &root->header);
 
     mutex_unlock(&root->lock);
@@ -406,9 +410,14 @@ void common_dmabuf_heap_release(struct heap_mem *heap_obj)
 {
 	struct heap_root *root = heap_obj->root;
 
-	mutex_lock(&root->lock);
-	kref_put(&heap_obj->refcount, __common_dmabuf_heap_release);
-	mutex_unlock(&root->lock);
+	if(!mutex_is_locked(&root->lock)) {
+		mutex_lock(&root->lock);
+		kref_put(&heap_obj->refcount, __common_dmabuf_heap_release);
+		mutex_unlock(&root->lock);
+	}
+	else {
+		kref_put(&heap_obj->refcount, __common_dmabuf_heap_release);
+	}
 }
 EXPORT_SYMBOL(common_dmabuf_heap_release);
 
@@ -530,6 +539,7 @@ struct heap_mem *common_dmabuf_heap_rsv_iova_map(struct heap_root *root, int fd,
 
 	kref_init(&heap_obj->refcount);
 
+	INIT_LIST_HEAD(&heap_obj->list);
 	list_add(&heap_obj->list, &root->header);
 
 	mutex_unlock(&root->lock);
@@ -576,9 +586,14 @@ void common_dmabuf_heap_rsv_iova_unmap(struct heap_mem *heap_obj)
 {
 	struct heap_root *root = heap_obj->root;
 
-	mutex_lock(&root->lock);
-	kref_put(&heap_obj->refcount, __common_dmabuf_heap_rsv_iova_unmap);
-	mutex_unlock(&root->lock);
+	if(!mutex_is_locked(&root->lock)) {
+		mutex_lock(&root->lock);
+		kref_put(&heap_obj->refcount, __common_dmabuf_heap_rsv_iova_unmap);
+		mutex_unlock(&root->lock);
+	}
+	else {
+		kref_put(&heap_obj->refcount, __common_dmabuf_heap_rsv_iova_unmap);
+	}
 }
 EXPORT_SYMBOL(common_dmabuf_heap_rsv_iova_unmap);
 
@@ -586,9 +601,11 @@ void common_dmabuf_heap_rsv_iova_uninit(struct heap_root *root)
 {
 	struct heap_mem *h, *tmp;
 
+	mutex_lock(&root->lock);
 	list_for_each_entry_safe(h, tmp, &root->header, list) {
 		common_dmabuf_heap_rsv_iova_unmap(h);
 	}
+	mutex_unlock(&root->lock);
 }
 EXPORT_SYMBOL(common_dmabuf_heap_rsv_iova_uninit);
 
-- 
2.53.0

